drivers: counter: Adapt to use device tree

nordic-krch opened the pull request:
Modifying counter drivers (rtc and timer) to rely completely on device tree and not on Kconfig of MDK flags.

Adapting dtsi for all SoCs and adapting test configuration.

nordic-krch committed:
drivers: counter: Adapt to use device tree

Modifying counter drivers (rtc and timer) to rely completely on
device tree and not on Kconfig of MDK flags.

Adapting dtsi for all SoCs and adapting test configuration.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: counter: maxim_ds3231: Remove unused test

Remove test which was copied from counter_basic_api and it was
always skipped.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
hal_nordic: Change scheme for RTC and TIMER reservation

In general, RTC and TIMER driver implements counter API but there
are exception when those peripherals are used in a custom way
(e.g. for system timer or bluetooth). In that case, system must
prevent using counter based on a reserved instance. Previously,
it was managed by Kconfig options but that cannot be maintained
when switching to devicetree configuration of the counter driver.

A new approach removes Kconfig options and instead adds static
asserts in the files which are using direct peripherals. Those
asserts check if given node is not enabled in the device tree.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

jciupis commented on modules/hal_nordic/nrf_802154/radio/platform/nrf_802154_random_zephyr.c:
```diff
@@ -8,6 +8,10 @@
 #include <zephyr/drivers/entropy.h>
 
+/* Ensure that counter driver for TIMER1 is not enabled. */
+BUILD_ASSERT(DT_NODE_HAS_STATUS(DT_NODELABEL(timer1), disabled),
```
I don't understand why this change is needed here. Could you please explain?

nordic-krch replied:
So there is a counter API which is generic API to be used for TIMER and RTC. The configuration for counter instances used to be split between Kconfig and DT so to use one instance you had to enable it in Kconfig.
There are cases when TIMER or RTC is used outside of counter API (system timer, radio driver or bluetooth) in that case we need to protect against using same instance twice, e.g. if radio driver is used TIMER1 cannot be used for counter.

We used to have Kconfig flags for that but after moving counter configuration to DT they are useless.

So assert checks if timer1 node is disabled when radio driver is used.

jciupis replied:
> if radio driver is used TIMER1 cannot be used for counter

That's reasonable, but in my opinion it's messy to put this check in a completely unrelated platform file that implements random API for the 802.15.4 driver. If anyone creates their own implementation of this API, for example using a different PRNG, they will lose this check. Wouldn't it be better to introduce this dependency in 802.15.4 driver's Kconfig by adding something like:
```
depends on $(dt_nodelabel_has_status,timer1,disabled)
```
or similar (not sure if the syntax I used is correct)?

jciupis replied:
LGTM, @e-rk what do you think?

e-rk replied:
My opinion is that having the check as @jciupis suggested would be preferable. That way we stay consistent with the rest of the system. The lack of information is the problem of Kconfig itself and I believe that it should be addressed there. If every module ended up with it's own dependency checking that is parallel to Kconfig we would end up in a mess.

nordic-krch replied:
Ok, i pushed take 2 where i used Kconfig approach. If there is a resource conflict then compilation fails without any verbal explanation.

anangl commented on drivers/counter/Kconfig.nrfx:
```diff
@@ -3,108 +3,39 @@
 
 config COUNTER_NRF_TIMER
 	bool
+	default y
```
nit
```suggestion
	def_bool y
```

anangl commented on drivers/counter/Kconfig.nrfx:
```diff
@@ -3,108 +3,39 @@
+		 $(dt_nodelabel_bool_prop,timer3,zli) || \
+		 $(dt_nodelabel_bool_prop,timer4,zli)
+	depends on COUNTER_NRF_RTC || COUNTER_NRF_TIMER
+	select ZERO_LATENCY_IRQS
```
I don't think this `select` is a good idea. Why not ensure instead with a build assertion in the driver that `ZERO_LATENCY_IRQS` is enabled when `zli` property is set for a given (enabled) instance?

nordic-krch replied:
Removed. If ZLI is not enabled but used there is already a BUILD_ASSERT.

anangl commented on drivers/counter/counter_nrfx_rtc.c:
```diff
@@ -670,72 +672,59 @@ static const struct counter_driver_api counter_nrfx_driver_api = {
 			.flags = COUNTER_CONFIG_INFO_COUNT_UP,		       \
-			.channels = RTC_PROP(idx, fixed_top) ?		       \
-				RTC##idx##_CC_NUM : RTC##idx##_CC_NUM - 1      \
+			.channels = DT_INST_PROP(idx, fixed_top) ?	       \
+				DT_INST_PROP(idx, cc_num) : DT_INST_PROP(idx, cc_num) - 1 \
```
How about:
```suggestion
			.channels = DT_INST_PROP(idx, fixed_top)	       \
				  ? DT_INST_PROP(idx, cc_num)		       \
				  : DT_INST_PROP(idx, cc_num) - 1	       \
```

anangl commented on samples/hello_world/src/main.c:
```diff
@@ -8,5 +8,6 @@
 
 void main(void)
 {
+	printk("%04x\n", __builtin_bswap16(0x1234));
```
Unintentional change.

anangl commented on drivers/counter/counter_nrfx_timer.c:
```diff
@@ -389,83 +391,62 @@ static const struct counter_driver_api counter_nrfx_driver_api = {
+	static const struct counter_nrfx_config nrfx_counter_##idx##_config = {		\
+		.info = {								\
+			.max_top_value = (DT_INST_PROP(idx, max_bit_width) == 32) ?	\
+					0xffffffff : 0x0000ffff,			\
```
Maybe `BIT64_MASK()` could be used for this?

anangl commented on drivers/counter/counter_nrfx_timer.c:
```diff
@@ -389,83 +391,62 @@ static const struct counter_driver_api counter_nrfx_driver_api = {
+			.max_top_value = (DT_INST_PROP(idx, max_bit_width) == 32) ?	\
+					0xffffffff : 0x0000ffff,			\
+			.freq = TIMER_CLOCK((NRF_TIMER_Type *)DT_INST_REG_ADDR(idx)) /	\
+					(1 << DT_INST_PROP(idx, prescaler)),		\
```
```suggestion
					BIT(DT_INST_PROP(idx, prescaler)),		\
```

nordic-krch committed:
boards: arm: nordic: Cleanup timers setup

All boards were enabling timer0-2 instances. Git history shows
that it is the result of copy-pasting. It is redundant thus removing.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

anangl commented on tests/drivers/counter/counter_basic_api/boards/nrf52840dk_nrf52811.overlay:
```diff
@@ -1,11 +1,14 @@
 &timer2 {
+	status = "okay";
 	prescaler = <4>;
 };
```
`rtc0` still missing here and in `nrf52dk_nrf52810.overlay`.

anangl commented on tests/drivers/counter/counter_basic_api/boards/nrf52840dk_nrf52840.overlay:
```diff
@@ -1,23 +1,34 @@
+
+&rtc2 {
+	status = "okay";
 	ppi-wrap;
```
Previously one instance was tested with `ppi-wrap` and one without. Shouldn't we keep this scenario?

nordic-krch committed:
hal_nordic: Add protection against resource conflict

Add check in CMake files to prevent resource overlap for
TIMER0, TIMER1, RTC0.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

anangl commented on subsys/bluetooth/controller/CMakeLists.txt:
```diff
@@ -20,6 +20,18 @@ zephyr_library_sources_ifdef(
   )
 
 if(CONFIG_BT_LL_SW_SPLIT)
+  dt_nodelabel(rtc0_node NODELABEL "rtc0")
```
This is nRF-specific so I think it should go to `ll_sw/nrf.cmake`.

anangl commented on tests/drivers/counter/counter_basic_api/boards/nrf52dk_nrf52810.overlay:
```diff
@@ -1,11 +1,14 @@
 &timer2 {
+	status = "okay";
 	prescaler = <4>;
 };
```
One last missing `rtc0`. :)

e-rk commented on modules/hal_nordic/nrf_802154/CMakeLists.txt:
```diff
@@ -5,6 +5,13 @@ zephyr_library_named(nrf-802154-platform)
+  dt_nodelabel(timer1_node NODELABEL "timer1")
+  dt_node_has_status(status_result PATH ${timer1_node} STATUS okay)
+  if (${status_result})
+	  message(FATAL_ERROR "Resource conflict. IEEE802.15.4 Radio driver requires TIMER1 but "
```
I might be wrong, but isn't there a policy that Kconfig/DTS errors must not result in CMake fatal errors, because this effectively locks-out the user from `menuconfig`?
On top of that, catching configuration errors in CMake doesn't seem right. Can it not be somehow achieved using the Kconfig alone?

e-rk reviewed (commented):
I have one comment. The Kconfigs `NRF_HW_RTCn_RESERVED` and `NRF_HW_TIMERn_RESERVED` were removed in favor of build errors in CMake/source files. One question from me:
Would it be possible to use different `compatible` for the `counter` driver and some other special purpose of the peripheral? For example, the RTC1 can be used as either `counter` or the system timer. By default, the DTS description would designate the RTC1 as `nordic,nrf-rtc-systemtimer` with default status `okay`. This would automatically configure the RTC1 to operate as system timer without conflicting with the counter, which would have a `compatible = "nordic,nrf-rtc-counter"` instead. Similar thing could be done for `TIMER` when it is used as counter or has special purpose in `nrf-802154`.

nordic-krch committed:
hal_nordic: Add protection against resource conflict (take 2)

Commit will be squeezed once agreement is reached.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>
